File: bindings/ruby/lib/proven/safe_uuid.rb
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
# frozen_string_literal: true

# Safe UUID operations via FFI to libproven.
# ALL computation delegates to Idris 2 compiled code.

module Proven
  module SafeUuid
    class << self
      # Generate a v4 (random) UUID string.
      # Returns nil on error.
      #
      # @return [String, nil] canonical UUID string (e.g., "550e8400-...")
      def v4
        # UUIDResult = { i32 status, [16]u8 uuid_bytes } = 20 bytes
        buf = Fiddle::Pointer.malloc(24, Fiddle::RUBY_FREE)

        fn = Fiddle::Function.new(
          FFI.handler["proven_uuid_v4"],
          [Fiddle::TYPE_VOIDP],
          Fiddle::TYPE_VOID,
          need_gvl: true
        )
        fn.call(buf)

        packed = buf.to_str(20)
        status = packed[0, 4].unpack1("l!")
        return nil unless status == FFI::STATUS_OK

        uuid_bytes = packed[4, 16]
        uuid_to_string(uuid_bytes)
      rescue Fiddle::DLError, TypeError
        nil
      end

      # Parse a UUID string, returning the canonical form.
      # Returns nil on invalid input.
      #
      # @param uuid_string [String]
      # @return [String, nil]
      def parse(uuid_string)
        return nil if uuid_string.nil?
        ptr, len = FFI.str_to_ptr(uuid_string)

        # UUIDResult = { i32 status, [16]u8 uuid } = 20 bytes
        buf = Fiddle::Pointer.malloc(24, Fiddle::RUBY_FREE)

        fn = Fiddle::Function.new(
          FFI.handler["proven_uuid_parse"],
          [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T],
          Fiddle::TYPE_VOID,
          need_gvl: true
        )
        fn.call(buf, ptr, len)

        packed = buf.to_str(20)
        status = packed[0, 4].unpack1("l!")
        return nil unless status == FFI::STATUS_OK

        uuid_bytes = packed[4, 16]
        uuid_to_string(uuid_bytes)
      rescue Fiddle::DLError, TypeError
        nil
      end

      # Check if string is valid UUID format.
      #
      # @param uuid_string [String]
      # @return [Boolean]
      def valid?(uuid_string)
        !parse(uuid_string).nil?
      end

      # Check if a UUID is the nil UUID (all zeros).
      #
      # @param uuid_string [String]
      # @return [Boolean, nil]
      def nil_uuid?(uuid_string)
        parsed = parse(uuid_string)
        return nil if parsed.nil?
        parsed == "00000000-0000-0000-0000-000000000000"
      end

      # Get the version of a parsed UUID.
      #
      # @param uuid_string [String]
      # @return [Integer, nil]
      def version(uuid_string)
        parsed_bytes = parse_to_bytes(uuid_string)
        return nil if parsed_bytes.nil?

        # Pack bytes and call proven_uuid_version
        packed_uuid = parsed_bytes.pack("C16")
        uuid_ptr = Fiddle::Pointer.to_ptr(packed_uuid)

        # UUID is passed as 16 bytes (struct by value), but since it's
        # exactly 16 bytes, on x86_64 it fits in two registers.
        # Use the int-based approach: pass as two i64 values.
        high = parsed_bytes[0, 8].pack("C8").unpack1("Q>")
        low = parsed_bytes[8, 8].pack("C8").unpack1("Q>")

        FFI.invoke_u8(
          "proven_uuid_version",
          [Fiddle::TYPE_LONG_LONG, Fiddle::TYPE_LONG_LONG],
          [high, low]
        )
      rescue Fiddle::DLError, TypeError
        nil
      end

      private

      # Format 16 bytes as canonical UUID string.
      def uuid_to_string(bytes)
        hex = bytes.unpack1("H*")
        "#{hex[0, 8]}-#{hex[8, 4]}-#{hex[12, 4]}-#{hex[16, 4]}-#{hex[20, 12]}"
      end

      # Parse UUID string into byte array (for passing to C).
      def parse_to_bytes(uuid_string)
        return nil if uuid_string.nil?
        return nil unless uuid_string.length == 36

        hex = uuid_string.delete("-")
        return nil unless hex.length == 32
        return nil unless hex.match?(/\A[0-9a-fA-F]+\z/)

        hex.scan(/../).map { |pair| pair.to_i(16) }
      end
    end
  end
end

File: bindings/php/src/Proven/SafeUUID.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeUUID - FFI wrapper for proven_uuid_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe UUID operations via libproven FFI.
 *
 * Every method calls the corresponding libproven FFI function.
 */
final class SafeUUID
{
    /**
     * Generate a new UUID v4.
     *
     * @return string|null The UUID as a string, or null on error.
     */
    public static function v4(): ?string
    {
        $result = FFI::getLib()->proven_uuid_v4();
        if ($result->status !== 0) {
            return null;
        }
        return self::uuidToString($result->uuid);
    }

    /**
     * Parse a UUID string into a UUID struct and return as a normalized string.
     *
     * @return string|null The normalized UUID string, or null on invalid input.
     */
    public static function parse(string $uuid): ?string
    {
        $result = FFI::getLib()->proven_uuid_parse($uuid, strlen($uuid));
        if ($result->status !== 0) {
            return null;
        }
        return self::uuidToString($result->uuid);
    }

    /**
     * Check if a UUID is the nil UUID (all zeros).
     *
     * @param string $uuid A UUID string to check.
     * @return bool|null True if nil, false if not, null on parse error.
     */
    public static function isNil(string $uuid): ?bool
    {
        $parsed = FFI::getLib()->proven_uuid_parse($uuid, strlen($uuid));
        if ($parsed->status !== 0) {
            return null;
        }
        return FFI::getLib()->proven_uuid_is_nil($parsed->uuid);
    }

    /**
     * Get the version number of a UUID.
     *
     * @param string $uuid A UUID string.
     * @return int|null The version (1-5), or null on parse error.
     */
    public static function version(string $uuid): ?int
    {
        $parsed = FFI::getLib()->proven_uuid_parse($uuid, strlen($uuid));
        if ($parsed->status !== 0) {
            return null;
        }
        return FFI::getLib()->proven_uuid_version($parsed->uuid);
    }

    /**
     * Convert a UUID struct to its string representation via FFI.
     */
    private static function uuidToString(mixed $uuid): ?string
    {
        $result = FFI::getLib()->proven_uuid_to_string($uuid);
        if ($result->status !== 0 || $result->value === null) {
            return null;
        }
        $str = \FFI::string($result->value, $result->length);
        FFI::getLib()->proven_free_string($result->value);
        return $str;
    }
}
